[FLINK-1436]  refactor CLiFrontend to provide more identifiable and meaningful error messages

mxm opened the pull request:
https://issues.apache.org/jira/browse/FLINK-1436

The main changes are:

* improve the error messages
* remove the verbose flag and make it always verbose
* do not show the full help text on errors.
* check for incomplete arguments

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -945,13 +948,10 @@ private int handleError(Throwable t) {
 	
 	/**
-	 * Parses the command line arguments and starts the requested action.
+	 * Parses the command lin e arguments and starts the requested action.
```
typo

mxm replied:
Thanks.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -960,14 +960,15 @@ public int parseParameters(String[] args) {
-			System.out.println("Please specify an action.");
 			printHelp();
+			System.out.println();
+			System.out.println("Please specify an action.");
```
instead of adding an additional call each time for the new-line, why aren't you pre-pending a '\n' to the string?

mxm replied:
Personally, I find it much easier to read than the semantically identical \n.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -413,29 +413,33 @@ protected int info(String[] args) {
 		PackagedProgram program;
 		try {
 			program = buildProgram(line);
+		// Java 7 supports } catch(FileNotFoundException e | ProgramInvocationException e2) {
```
I think that comment is not necesarry

mxm replied:
Yep, will remove.

rmetzger commented:
Testing JIRA integration FLINK-1436

Lets see

StephanEwen commented:
Looks good to me. 

+1 to merge (with commit squashing)...

rmetzger commented:
@mxm: Since you have some experience with the CliFrontend now, would you mind looking into https://issues.apache.org/jira/browse/FLINK-1424 as well?

mxm commented:
@rmetzger @StephanEwen I squashed the commits, merged the recent changes in the CliFrontend on the master, and addressed [FLINK-1424](https://issues.apache.org/jira/browse/FLINK-1424).

rmetzger commented:
I think it would be better not to print the help if the user specified something incorrectly. Maybe just the error message and a note that -h prints the help?

I've tried out the change, but now, the message is as the very bottom of the output. Its now probably even harder to find it.

**Bad** (see below for *Good*)

```
robert@robert-da ~/flink-workdir/flink2/build-target (git)-[flink-1436] % ./bin/flink ./examples/flink-java-examples-0.9-SNAPSHOT-WordCount.jar 

Action "run" compiles and runs a program.

  Syntax: run [OPTIONS] <jar-file> <arguments>
  "run" action options:
     -c,--class <classname>           Class with the program entry point ("main"
                                      method or "getPlan()" method. Only needed
                                      if the JAR file does not specify the class
                                      in its manifest.
     -m,--jobmanager <host:port>      Address of the JobManager (master) to
                                      which to connect. Specify 'yarn-cluster'
                                      as the JobManager to deploy a YARN cluster
                                      for the job. Use this flag to connect to a
                                      different JobManager than the one
                                      specified in the configuration.
     -p,--parallelism <parallelism>   The parallelism with which to run the
                                      program. Optional flag to override the
                                      default value specified in the
                                      configuration.
  Additional arguments if -m yarn-cluster is set:
     -yD <arg>                            Dynamic properties
     -yj,--yarnjar <arg>                  Path to Flink jar file
     -yjm,--yarnjobManagerMemory <arg>    Memory for JobManager Container [in
                                          MB]
     -yn,--yarncontainer <arg>            Number of YARN container to allocate
                                          (=Number of Task Managers)
     -yq,--yarnquery                      Display available YARN resources
                                          (memory, cores)
     -yqu,--yarnqueue <arg>               Specify YARN queue.
     -ys,--yarnslots <arg>                Number of slots per TaskManager
     -yt,--yarnship <arg>                 Ship files in the specified directory
                                          (t for transfer)
     -ytm,--yarntaskManagerMemory <arg>   Memory per TaskManager Container [in
                                          MB]

Invalid action: "./examples/flink-java-examples-0.9-SNAPSHOT-WordCount.jar"
1 robert@robert-da ~/flink-workdir/flink2/build-target (git)-[flink-1436]
```

The "info" command is over-engineered in my optionion. It contains only one possible option, which is "-e" for execution plan. I would vote to remove the info action and call it "plan" or so. 
Or keep its "info" name and print the plan by default (this is not @mxm's fault .. but it would be nice to fix this with the PR)
```
 ./bin/flink info  ./examples/flink-java-examples-0.9-SNAPSHOT-WordCount.jar  

Action "info" displays information about a program.

  Syntax: info [OPTIONS] <jar-file> <arguments>
  "info" action options:
     -c,--class <classname>           Class with the program entry point ("main"
                                      method or "getPlan()" method. Only needed
                                      if the JAR file does not specify the class
                                      in its manifest.
     -e,--executionplan               Show optimized execution plan of the
                                      program (JSON)
     -m,--jobmanager <host:port>      Address of the JobManager (master) to
                                      which to connect. Specify 'yarn-cluster'
                                      as the JobManager to deploy a YARN cluster
                                      for the job. Use this flag to connect to a
                                      different JobManager than the one
                                      specified in the configuration.
     -p,--parallelism <parallelism>   The parallelism with which to run the
                                      program. Optional flag to override the
                                      default value specified in the
                                      configuration.

Error: Specify one of the above options to display information.
```

**Good**

What I liked was the error reporting when passing an invalid file as the jar file:
```
robert@robert-da ~/flink-workdir/flink2/build-target (git)-[flink-1436] % ./bin/flink run ../flink-yarn-0.9-SNAPSHOT/LICENSE

Action "run" compiles and runs a program.

  Syntax: run [OPTIONS] <jar-file> <arguments>
  "run" action options:
     -c,--class <classname>           Class with the program entry point ("main"
                                      method or "getPlan()" method. Only needed
                                      if the JAR file does not specify the class
                                      in its manifest.
     -m,--jobmanager <host:port>      Address of the JobManager (master) to
                                      which to connect. Specify 'yarn-cluster'
                                      as the JobManager to deploy a YARN cluster
                                      for the job. Use this flag to connect to a
                                      different JobManager than the one
                                      specified in the configuration.
     -p,--parallelism <parallelism>   The parallelism with which to run the
                                      program. Optional flag to override the
                                      default value specified in the
                                      configuration.
  Additional arguments if -m yarn-cluster is set:
     -yD <arg>                            Dynamic properties
     -yj,--yarnjar <arg>                  Path to Flink jar file
     -yjm,--yarnjobManagerMemory <arg>    Memory for JobManager Container [in
                                          MB]
     -yn,--yarncontainer <arg>            Number of YARN container to allocate
                                          (=Number of Task Managers)
     -yq,--yarnquery                      Display available YARN resources
                                          (memory, cores)
     -yqu,--yarnqueue <arg>               Specify YARN queue.
     -ys,--yarnslots <arg>                Number of slots per TaskManager
     -yt,--yarnship <arg>                 Ship files in the specified directory
                                          (t for transfer)
     -ytm,--yarntaskManagerMemory <arg>   Memory per TaskManager Container [in
                                          MB]

org.apache.flink.client.program.ProgramInvocationException: Error while opening jar file '../flink-yarn-0.9-SNAPSHOT/LICENSE'. error in opening zip file
	at org.apache.flink.client.program.PackagedProgram.getEntryPointClassNameFromJar(PackagedProgram.java:466)
	at org.apache.flink.client.program.PackagedProgram.<init>(PackagedProgram.java:134)
	at org.apache.flink.client.program.PackagedProgram.<init>(PackagedProgram.java:102)
	at org.apache.flink.client.CliFrontend.buildProgram(CliFrontend.java:731)
	at org.apache.flink.client.CliFrontend.run(CliFrontend.java:332)
	at org.apache.flink.client.CliFrontend.parseParameters(CliFrontend.java:1078)
	at org.apache.flink.client.CliFrontend.main(CliFrontend.java:1102)
Caused by: java.util.zip.ZipException: error in opening zip file
	at java.util.zip.ZipFile.open(Native Method)
	at java.util.zip.ZipFile.<init>(ZipFile.java:215)
	at java.util.zip.ZipFile.<init>(ZipFile.java:145)
	at java.util.jar.JarFile.<init>(JarFile.java:154)
	at java.util.jar.JarFile.<init>(JarFile.java:118)
	at org.apache.flink.client.program.PackagedProgram.getEntryPointClassNameFromJar(PackagedProgram.java:464)
	... 6 more

The exception above occurred while trying to run your command.
```

mxm commented:
Thanks for your feedback, @rmetzger . The error message is at the bottom because that way it is most easily identifiable by the user (no scrolling necessary). Before, we printed the error and then the help which let the help shadow the error message.

I changed the error reporting in case the user didn't specify an action.

Concerning the printing of the help message, you're probably right. Let's just print the help if the user asks for it. Now it prints:
> "./examples/flink-java-examples-0.9-SNAPSHOT-WordCount.jar" is not a valid action.
> Valid actions are "run", "list", "info", or "cancel".

Additionally, let's 
* change `info` to print the plan by default
* change `cancel` to accept the job id as a parameter instead of an option
* change `list` to print scheduled and running jobs by default

rmetzger commented:
Cool, thank you. I'll try it out as soon as I find time.

mxm commented:
Any further comments on the changes?

StephanEwen commented:
Looks like a good improvement to me. 

+1 to merge

rmetzger commented:
Good work. Thank you.

I think its time for your first push to master ;)

rmetzger commented:
Remember to include "This closes #311" ;)

mxm commented:
@rmetzger Thanks but I think it should be "This closes #331" :)
